### PR TITLE
Removing int8 column option from parquet byte_array writing

### LINKED ISSUE #11536: Need to remove int8 column support in parquet binary code path.
"Does Spark differentiate binary types stored as int8 and uint8? Are both of these equivalent from Spark’s perspective? Can Spark normalize inputs to use only one or the other? It feels very suspicious to me that we’ve decided to accept either one across the binary code paths."

__Originally posted by @bdice in https://github.com/rapidsai/cudf/pull/11526#discussion_r945891148__

It was pointed out that the code path in parquet support both INT8 columns and UINT8 columns. We should simply be opinionated here and after discussion it was decided to use UINT8.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
## Description
As suggested in #11526 and captured in issue #11536 the usage of both INT8 and UINT8 as supported types for byte_arrays is unnecessary and adds complexity to the code. This change removes INT8 as an option and only allows UINT8 columns to be written out as byte_arrays. ~~This matches with cudf string columns which contain an INT8 column for data.~~

closes #11536

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/io/parquet_test.cpp
Comment (bdice) [inline cpp/tests/io/parquet_test.cpp:4218]: Match the type (`int8_t`) here. However, the bigger question is whether this test still fits the criteria explained in the comment. It appears it does not exceed the maximum UTF-8 value anymore.

File: cpp/src/lists/dremel.cu
Comment (bdice) [inline cpp/src/lists/dremel.cu:195]: Check the type second, to match the style of the rest of the code.
```suggestion
      if (output_as_byte_array && child.type().id() == type_id::INT8) {
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: cpp/tests/io/parquet_test.cpp
Comment (etseidl) [reply cpp/tests/io/parquet_test.cpp:4218]: I wouldn't change this test.  Binary in parquet is unsigned, so you still want to test that `0xfe` doesn't wind up as `0xf7bfbfbf` in the statistics if it's written as binary and not string.  This should have nothing to do with how binary is represented internally in cuDF columns.

**Review by upsj — APPROVED** [review]
LGTM except for the aforementioned concerns about values outside int8's range.

**Comment by davidwendt** [comment]
Based on the description, I would've expected this would strictly used UINT8 instead of INT8 since the sign bit has no meaning in plain binary data like this. Also, it would probably help with the boundary issues mentioned in the other review comments as well.
I think the consistency with strings argument is overly emphasized and should really not be a consideration in this implementation.

File: cpp/tests/io/parquet_test.cpp
Comment (hyperbolic2346) [reply cpp/tests/io/parquet_test.cpp:4218]: Thank you @etseidl, It is important to remember the origin of these values. They were selected specifically to ensure that the stats were not for a string by being one larger than the max UTF-8 value.

> Match the type (int8_t) here.

I originally was doing that, but as Ed mentions below, binary stats in parquet are unsigned and I am comparing to the actual stats values.

**Comment by hyperbolic2346** [comment]
> I think the consistency with strings argument is overly emphasized and should really not be a consideration in this implementation.

I would typically use unsigned int values for bytes as well and I would agree with you about the string argument if it weren't for the strong coupling between the two that exists. When reading binary, we read it as a string always and can decompose the column and built it again as a list of bytes if desired. The column type that we get from pulling apart a string column is an int8 column, for reasons unknown to me. This seemed to imply a precedence for int8 byte representations inside cudf and also made the easy path for conversion to be list<int8>. As a result, reading binary data from a parquet file returns a list<int8> type column right now.

That said, the easy path doesn't mean that it is the best path by a long shot. If we have consensus that uint8 should be the type we use and expect, I am ok with that, but it will require more than a few changes. It is certainly more natural to fit into the parquet statistics.

**Comment by etseidl** [comment]
> The column type that we get from pulling apart a string column is an int8 column, for reasons unknown to me.

Beyond scope for this PR, but I'd be interested in a discussion about this.  Since strings in cuDF are UTF-8, not c-style char arrays, why not use UINT8 as the child column type for both binary and string data?  Is it for ease of binding to python strings?  `string_view` has to cast to `unsigned char *` to compare two strings already. Then again, casts are cheap and refactoring expenssive, so 🤷

**Comment by hyperbolic2346** [comment]
>  why not use UINT8 as the child column type for both binary and string data

From what I can gather, when strings were implemented cudf didn't have a uint8 type. I agree that unsigned is a better option, but I wonder the impact that would have externally by users. I think the best options is to create an issue to discuss changing the behavior.

**Comment by davidwendt** [comment]
I don't think consistency with a different type (i.e. `string_view`) is a compelling reason to use INT8 here. The `string_view` has its own requirements and dependencies (and things that depend on it). The design and implementation of `string_view` can be discussed outside the scope of this PR. Regardless, I see no reason to prevent a byte-array to use UINT8.

File: cpp/tests/io/parquet_test.cpp
Comment (ttnghia) [inline cpp/tests/io/parquet_test.cpp:4230]: I don't think we will need `static_cast`. In case of potential over/underflow, we should use `reinterpret_cast`.

Comment (bdice) [reply cpp/tests/io/parquet_test.cpp:4230]: @ttnghia Can you explain why you think that `reinterpret_cast` is the correct choice here? We are relying on UB in the `static_cast`, but I don't think `reinterpret_cast` makes this better (relies on UB for type aliasing instead of UB for overflow). I don't see this as a case listed in "Only the following conversions can be done with `reinterpret_cast`: https://en.cppreference.com/w/cpp/language/reinterpret_cast

@hyperbolic2346 I think this is an example that adds to the evidence that we may want to use uint8 instead of int8.

Comment (ttnghia) [reply cpp/tests/io/parquet_test.cpp:4230]: `static_cast` may do the conversion which may incur rounding off if the input value are out of bound of the target type.

`reinterpret_cast` on the other hand will not do anything. It just treats the input value as it is of the target type so there will not be any rounding off.

Comment (ttnghia) [reply cpp/tests/io/parquet_test.cpp:4230]: Since we want the input values to be copied exactly into memory without data corruption, `reinterpret_cast` should be the right tool.

Comment (bdice) [reply cpp/tests/io/parquet_test.cpp:4230]: @ttnghia The suggestion of `reinterpret_cast` doesn't compile. https://godbolt.org/z/G7r6a5x86

`reinterpret_cast` is typically only used for pointers/references. See also: https://stackoverflow.com/questions/18544849/how-to-use-reinterpret-cast-in-c

Comment (ttnghia) [reply cpp/tests/io/parquet_test.cpp:4230]: Oh no.....
So there would be another cleaner way than using multiple `static_cast`: construct a `lists_column_wrapper<uint8_t>` like previously, then call `cudf::bit_cast` to `lists_column_wrapper<int8_t>`.

Comment (ttnghia) [reply cpp/tests/io/parquet_test.cpp:4230]: But keeping the current `static_cast` way is not a big deal anyway :smile:

Comment (hyperbolic2346) [reply cpp/tests/io/parquet_test.cpp]: I believe this turned into a non-issue. Feel free to bring it back if it isn't resolved with the latest code.

**Comment by hyperbolic2346** [comment]
Updated this to change the column type to uint8 instead of int8. Now binary writes only accept uint8 columns and binary reads come back as uint8 columns.

**Comment by revans2** [comment]
Can we please hold off on merging this until the Spark Rapids plugin can adjust the plugin to deal with these changes?  Otherwise we are going to broken and will have to scramble to get a fix in.

**Comment by hyperbolic2346** [comment]
> Can we please hold off on merging this until the Spark Rapids plugin can adjust the plugin to deal with these changes? Otherwise we are going to broken and will have to scramble to get a fix in.

Absolutely, but I think we have to be sure to finish the review process here so that we end up with something that is approved before Spark Rapids makes changes to match. So to the reviewers, please continue to review this code and approve if warranted.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/utilities/column_utils.cuh
Comment (davidwendt) [inline cpp/src/io/utilities/column_utils.cuh:66]: Just curious why this `mutable` is needed? I've not seen this in a device lambda.

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/utilities/column_buffer.cpp
Comment (mythrocks) [inline cpp/src/io/utilities/column_buffer.cpp:94]: It seems like this is the bit that's confounding `spark-rapids`. In doing this, the metadata section in the output goes from:
```
FileMetaData{schema: message schema {
  required binary mybin;
}
```
... to this:
```
FileMetaData{schema: message schema {
  required group mybin (LIST) {
    repeated group list {
      required int32 element (INTEGER(8,true));
    }
  }
}
```
The relabeling is breaking Spark's data schema inference logic.

**Comment by mythrocks** [comment]
The sticking point in this PR (for me) is [this change](https://github.com/rapidsai/cudf/pull/11539/files#diff-9def7814b3da2ba309710ff0550efefe8c3ea0b275ca6953ba72600c986deb9b) made to `column_buffer.cpp`. Here, a binary column is translated to the equivalent `UINT8` column (alongside a corresponding metadata change).

Consider a hypothetical unit-test in Apache Parquet that writes out a `BINARY` column, and reads it back to confirm that it is indeed `BINARY`. The expected file footer entry for it would be:
```
FileMetaData{schema: message schema {
  required binary mycol;
}
```
Such a test would fail with this change, because the column read back would be `LIST<INT8>`:
```
FileMetaData{schema: message schema {
  required group mycol (LIST) {
    repeated group list {
      required int32 element (INTEGER(8,true));
    }
  }
}
```
To wit, reading this back into Spark fails in my tests, because the read schema does not match the write schema. 
Would it be viable to revert the column conversion in `column_buffer.cpp`?

**Comment by ttnghia** [comment]
Is this still targeting to 22.10?

**Comment by hyperbolic2346** [comment]
No, this is a 22.12 feature. When I switched it, there wasn't a branch available yet. I'll update this.

**Comment by rwlee** [comment]
Sorry, updated the PR branch and didn't realize this was still targeted at 22.10

**Comment by rwlee** [comment]
Also resolves https://github.com/NVIDIA/spark-rapids/issues/6408 with the added changes that standardize casting to list of bytes. There should be no additional changes needed on the plugin side based on my testing + integration tests.

File: cpp/src/io/utilities/column_utils.cuh
Comment (rwlee) [reply cpp/src/io/utilities/column_utils.cuh]: It is not needed, I just removed it and ran some tests without `mutable` to confirm.

File: cpp/src/io/utilities/column_buffer.cpp
Comment (mythrocks) [reply cpp/src/io/utilities/column_buffer.cpp:94]: The way I tested this before was to write out `STRING` data to Parquet as `BINARY`, and then attempt to read it back. The corresponding Spark code was:
```scala
val filepath = "/tmp/strbin.parquet"
sql(" SELECT CAST('Hello, World' AS BINARY) AS strbin ").write.parquet(filepath)
spark.read.parquet(filepath).show
```

Have we confirmed that this works, after this change?

Comment (mythrocks) [reply cpp/src/io/utilities/column_buffer.cpp:94]: > This was (surpisingly) not the issue causing the schema inconsistencies -- that was an implicit call on the spark side to a cast to list of binary that was not doing this kind of uint8 conversion. 

In my example above, if the user casts explicitly to `BINARY` before writing to Parquet, then they expect to read `BINARY` when the output file is read. If it reads back as `LIST<BYTE>`, I don't think the change is right.

File: cpp/src/reshape/byte_cast.cu
Comment (rwlee) [reply cpp/src/reshape/byte_cast.cu:113]: I had the same issue, I was repurposing the logic used in other portions of the code but the double release isn't intuitive to me still.

File: cpp/src/io/utilities/column_buffer.cpp
Comment (mythrocks) [reply cpp/src/io/utilities/column_buffer.cpp:94]: >  If it reads back as `LIST<BYTE>`, I don't think the change is right.

We just tested this. The column is written correctly as `BINARY`. This should appease Spark.

**Review by mythrocks — APPROVED** [review]

File: cpp/src/reshape/byte_cast.cu
Comment (bdice) [inline cpp/src/reshape/byte_cast.cu:113]: Should we fetch the null count before releasing the data, and pass it here? We should try to retain any information that may have been previously computed.

Generally, I think we're going to move away from lazily computing null counts as we work on adding streams to the public API. There are some hiccups with the lazy approach and streams, but primarily we aren't being as efficient as possible wrt avoiding calls to the `null_count` kernel because we aren't tracking information that we already may have computed once earlier.

**Review by bdice — APPROVED** [review]
Approving with one question. Happy to defer on that if you don't want to make those changes in this PR.

**Comment by rwlee** [comment]
Adding do not merge because of some upmerge issues. Upmerged to resolve some python errors and tested locally successfully... Continuing to debug but i don't want this merged accidently.

File: cpp/src/reshape/byte_cast.cu
Comment (rwlee) [reply cpp/src/reshape/byte_cast.cu:113]: I like the idea of moving away form lazily computed null counts and am happy to change here if we want to start now. In this case, it made sense for me to defer the null counting because byte cast is often use as an intermediate type for reading, writing, and hashing. Those are usually indifferent to null count and it made sense to keep it lazily computed while we still supported it.

**Comment by bdice** [comment]
rerun tests

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**
